devito/passes/iet/mpi.py

```diff
@@ -59,120 +59,116 @@ def _hoist_halospots(iet):
 
-    def rule1(dep, candidates, loc_dims):
-        # A reduction isn't a stopper to hoisting
-        return dep.write is not None and dep.write.is_reduction
+    Example:
+                                   haloupd v[t0]
+    for time                       for time
+      W v[t1]- R v[t0]             W v[t1]- R v[t0]
+      haloupd v[t1]                  haloupd v[t1]
+      R v[t1]                        R v[t1]
+      haloupd v[t0]                  R v[t0]
+      R v[t0]
 
-    rules = [rule0, rule1]
+    """
 
     # Precompute scopes to save time
     scopes = {i: Scope([e.expr for e in v]) for i, v in MapNodes().visit(iet).items()}
 
+    cond_mapper = _create_cond_mapper(iet)
+
     # Analysis
     hsmapper = {}
     imapper = defaultdict(list)
-    for iters, halo_spots in MapNodes(Iteration, HaloSpot, 'groupby').visit(iet).items():
-        for hs in halo_spots:
-            hsmapper[hs] = hs.halo_scheme
 
-            for f, v in hs.fmapper.items():
-                loc_dims = frozenset().union([q for d in v.loc_indices
-                                              for q in d._defines])
+    # Look for parent Iterations of children HaloSpots
+    for iters, halo_spots in MapNodes(Iteration, HaloSpot, 'groupby').visit(iet).items():
+        for i, hs0 in enumerate(halo_spots):
 
-                for n, i in enumerate(iters):
-                    if i not in scopes:
-                        continue
+            # Nothing to do if the HaloSpot is void
+            if hs0.halo_scheme.is_void:
+                continue
 
-                    candidates = [i.dim._defines for i in iters[n:]]
+            for hs1 in halo_spots[i+1:]:
+                # If there are Conditionals involved, both `hs0` and `hs1` must be
+                # within the same Conditional, otherwise we would break the control
+                if cond_mapper.get(hs0) != cond_mapper.get(hs1):
+                    continue
 
-                    all_candidates = set().union(*candidates)
-                    reads = scopes[i].getreads(f)
-                    if any(set(a.ispace.dimensions) & all_candidates
-                           for a in reads):
-                        continue
+                # If there are overlapping time accesses, skip
+                if any(i in hs0.halo_scheme.loc_values
+                        for i in hs1.halo_scheme.loc_values):
+                    continue
 
-                    for dep in scopes[i].d_flow.project(f):
-                        if not any(r(dep, candidates, loc_dims) for r in rules):
-                            break
-                    else:
-                        hsmapper[hs] = hsmapper[hs].drop(f)
-                        imapper[i].append(hs.halo_scheme.project(f))
-                        break
+                # Compare hs0 to subsequent halo_spots, looking for optimization
+                # possibilities
+                _process_halo_to_halo(hs0, hs1, iters, scopes, hsmapper, imapper)
```

Comment:
a function call here , but it smells like a disguise to split a very very long code block into two

Reply:
in retrospect, now that I've finished reviewing, we should probably:

* simplify (as much as possible!) what's currently inside `_process_halo_to_halo`
* inline it here directly

Reply:
Done